apps/prairielearn/src/lib/submissions-upload.ts

```diff
@@ -300,3 +242,368 @@ export async function uploadSubmissions(
+        sql.insert_variant,
+        {
+          course_id,
+          course_instance_id,
+          instance_question_id,
+          question_id: question.id,
+          authn_user_id: user.user_id,
+          user_id: user.user_id,
+          seed: row.Seed,
+          // Despite the fact that these values could change over the course of multiple
+          // submissions, we'll just use the first set of values we encounter. This
+          // is good enough for our purposes.
+          params: row.Params,
+          true_answer: row['True answer'],
+          options: row.Options,
+          number: row.Variant,
+        },
+        IdSchema,
+      ),
+  );
+
+  const submission_id = await sqldb.queryRow(
+    sql.insert_submission,
+    {
+      variant_id,
+      authn_user_id,
+      submitted_answer: row['Submitted answer'],
+      params: row.Params,
+      true_answer: row['True answer'],
+      submission_date: row['Submission date'],
+    },
+    IdSchema,
+  );
+
+  let manual_rubric_data: InstanceQuestionScoreInput['manual_rubric_data'] | null = null;
+
+  if (assessmentQuestion.manual_rubric_id) {
+    const rubric_items = await sqldb.queryRows(
+      sql.select_rubric_items,
+      { rubric_id: assessmentQuestion.manual_rubric_id },
+      RubricItemSchema,
+    );
+
+    await sqldb.execute(sql.update_assessment_instance_max_points, {
+      assessment_instance_id,
+      max_points: maxPoints,
+    });
+
+    // This is a best-effort process: we attempt to match uploaded rubric items to an
+    // existing rubric item in the database. If no match is found, the uploaded item
+    // is ignored and not applied.
+    const applied_rubric_item_ids: string[] = [];
+
+    if (row['Rubric Grading']?.items) {
+      for (const { description } of row['Rubric Grading'].items) {
+        const rubric_item = rubric_items.find((ri) => ri.description === description);
+        if (!rubric_item) {
+          continue;
+        }
+        applied_rubric_item_ids.push(rubric_item.id);
+      }
+    }
+    manual_rubric_data = {
+      rubric_id: assessmentQuestion.manual_rubric_id,
+      applied_rubric_items: applied_rubric_item_ids.map((id) => ({
+        rubric_item_id: id,
+      })),
+      adjust_points: row['Rubric Grading']?.adjust_points ?? null,
+    };
+  }
+
+  await updateInstanceQuestionScore(
+    assessment_id,
+    instance_question_id,
+    submission_id,
+    null,
+    {
+      manual_score_perc: null,
+      manual_points: run(() => {
+        if (assessmentQuestion.manual_rubric_id) {
+          return row['Rubric Grading']?.computed_points ?? null;
+        } else {
+          return row['Manual points'];
+        }
+      }),
+      auto_score_perc: null,
+      auto_points: row['Auto points'] ?? null,
+      feedback: row.Feedback,
+      manual_rubric_data,
+    },
+    authn_user_id,
+  );
+}
+
+async function processGroupSubmissionRow(record: any, context: ProcessingContext): Promise<void> {
```

Comment:
There seems to be a lot of duplication between these two functions. Can we collapse them into one, and maybe break out the individual vs. group behavior into their own smaller functions that we can branch to?